app/src/main/java/com/eis/phoneringer/AppManager.java

```diff
@@ -0,0 +1,96 @@
+package com.eis.phoneringer.structure;
+
+import android.content.Context;
+import android.media.Ringtone;
+import android.os.Build;
+import android.os.Handler;
+import android.os.Looper;
+import android.widget.Toast;
+
+import androidx.annotation.RequiresApi;
+
+import com.eis.smslibrary.SMSHandler;
+import com.eis.smslibrary.exceptions.InvalidSMSMessageException;
+import com.eis.smslibrary.exceptions.InvalidTelephoneNumberException;
+
+/**
+ * This is a singleton class used to manage a received RingCommand or to send one
+ *
+ * @author Alberto Ursino, Luca Crema, Alessandra Tonin, Marco Mariotto
+ */
+public class AppManager {
+
+    private static final String WRONG_PASSWORD_TOAST = "Wrong password";
+    private static final int TIMEOUT_TIME = 30 * 1000; //30 seconds
+
+    /**
+     * Instance of the class that is instantiated in getInstance method
+     */
+    private static AppManager instance = null;
+
+    /**
+     * Private constructor
+     */
+    private AppManager() {
+    }
+
+    /**
+     * @return the AppManager instance
+     */
+    public static AppManager getInstance() {
+        if (instance == null)
+            instance = new AppManager();
+        return instance;
+    }
+
+    /**
+     * If the password of the message received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand a ring command not null
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public void onRingCommandReceived(Context context, RingCommand ringCommand, final Ringtone ringtone) {
+        final RingtoneHandler ringtoneHandler = RingtoneHandler.getInstance();
+
+        if (checkPassword(context, ringCommand)) {
+            ringtoneHandler.playRingtone(ringtone);
+            //Timer: the ringtone is playing for TIMEOUT_TIME seconds.
+            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    ringtoneHandler.stopRingtone(ringtone);
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            Toast.makeText(context, WRONG_PASSWORD_TOAST, Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    /**
+     * Method used to send a RingCommand via SMS using the library class {@link SMSHandler}
+     *
+     * @param context     of the application
+     * @param ringCommand to send
+     * @throws InvalidTelephoneNumberException
+     * @throws InvalidSMSMessageException
```

Comment:
Are these two exceptions actually thrown? `parseCommand` doesn't throw anything, even though it probably should because we don't check `SMSPeer` validity in RingCommand

Reply:
You should probably go and edit RingCommandHandler.parseCommand and add `throws InvalidTelephoneNumberException, InvalidSMSMessageException`

Reply:
These exceptions can only be launched in this method because the SMSHandler is invoked only within it.

Reply:
`SMSHandler `doesn't throw any exception. Those who throw those two exceptions are `SMSPeer `and `SMSMessage`, which are created inside `RingCommandHandler.parseCommand`, so this method should throw them.

Reply:
Ops, you're right.
Is in the _ringCommandhandler_ class, when the SMSMessage is created, which can be launched.

Reply:
Well we have to update the RingCommandHandler class.